# Combining and Exporting Data

#### This Script Contains the Following Points:
#### 1. Import the data sets into Jupyter
#### 2. Check the dimensions of the imported dataframes
#### 3. Merge the dataframes
#### 4. Use the indicator argument to check whether there was a full match between the two dataframes
#### 5. Export data in pickle format

### 1. Import the datasets to Jupyter

In [5]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [6]:
#importing data
path = r'/Users/kimkmiz/Documents/Instacart Basket Analysis 2024'

In [7]:
#import new dataset as Orders_products_prior
orders_products_pr = df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'IC24 Original Data', 'orders_products_prior.csv'), index_col = False)

In [8]:
# Import dataset orders_clean.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'IC24 Prepared Data', 'orders_clean.csv'), index_col = False)

### 2. Check the dimensions of the imported dataframes

In [10]:
# Check the output
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [11]:
df_ords_prior.shape

(32434489, 4)

In [12]:
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id eval_set  order_number  \
0             0           0   2539329        1    prior             1   
1             1           1   2398795        1    prior             2   
2             2           2    473747        1    prior             3   
3             3           3   2254736        1    prior             4   
4             4           4    431534        1    prior             5   

   order_dow  order_hour_of_day  days_since_prior_order  
0          2                  8                     NaN  
1          3                  7                    15.0  
2          3                 12                    21.0  
3          4                  7                    29.0  
4          4                 15                    28.0

In [13]:
df_ords.shape

(3421083, 9)

**Observations on dataset dimensions**
- df_ords_prior is longer than df_ords
- df_ords is wider than df_ords_prior
- the columns eval_set, unnamed 0.1, and unnamed 0 are not necessary and should be dropped from df_ords
- order_dow should have been renamed order_day_of_week (need to further clean this)
- column could be added to df_ords to indicate if it is a first order or not based on missing value 

In [15]:
# Drop unnecessary columns
df_ords = df_ords.drop(['Unnamed: 0.1', 'Unnamed: 0', 'eval_set'], axis=1)

In [16]:
# Check the output
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [17]:
#rename order_dow to order_day_of_week for df_ords
df_ords.rename(columns={'order_dow': 'order_day_of_week'}, inplace=True)

In [18]:
# Check the output
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### 3. Merge the dataframes

In [20]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [21]:
#check the output
df_merged_large.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

### 4. Use the indicator argument to check if full match between dataframes

In [23]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

**Observations** 
- There are 32,434,489 rows in the merged dataframe
- Inner join with information from both datasets

### 5. Export in Pickle format

In [26]:
# Export data to pkl
df_merged_large.to_pickle(os.path.join(path, '02 Data','IC24 Prepared Data', 'orders_products_combined.pkl'))

In [27]:
# Export modified df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','IC24 Prepared Data', 'orders_final.csv'))